cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+        match (self, other) {
+            (
+                Fact::ValueMax {
+                    bit_width: bw_lhs,
+                    max: lhs,
+                },
+                Fact::ValueMax {
+                    bit_width: bw_rhs,
+                    max: rhs,
+                },
+            ) if bw_lhs == bw_rhs && add_width >= *bw_lhs => {
+                let computed_max = lhs.checked_add(*rhs)?;
+                Some(Fact::ValueMax {
+                    bit_width: *bw_lhs,
+                    max: computed_max,
+                })
+            }
+
+            (
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+                Fact::PointsTo { region },
+            )
+            | (
+                Fact::PointsTo { region },
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+            ) if *bw_max >= pointer_width && add_width >= *bw_max => {
+                let computed_region = MemoryRegion {
+                    max: region.max.checked_sub(*max)?,
+                };
+                Some(Fact::PointsTo {
+                    region: computed_region,
+                })
+            }
+
+            _ => None,
+        }
+    }
+
+    /// Computes the `uextend` of a value with the given facts.
+    pub fn uextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and we
+            // are filling zeroes into from_bits..to_bits, and
+            // bit_width and from_bits are exactly contiguous, then we
+            // have defined values in 0..to_bits (and because this is
+            // a zero-extend, the max value is the same).
+            Fact::ValueMax { bit_width, max } if *bit_width == from_width => Some(Fact::ValueMax {
+                bit_width: to_width,
+                max: *max,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Computes the `sextend` of a value with the given facts.
+    pub fn sextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and
+            // the MSB w.r.t. `from_width` is *not* set, then we can
+            // do the same as `uextend`.
+            Fact::ValueMax { bit_width, max }
+                if *bit_width == from_width && (*max & (1 << (*bit_width - 1)) == 0) =>
+            {
+                self.uextend(from_width, to_width)
+            }
+            _ => None,
+        }
+    }
+
+    /// Scales a value with a fact by a known constant.
+    pub fn scale(&self, width: u16, factor: u32) -> Option<Self> {
+        match self {
+            Fact::ValueMax { bit_width, max } if *bit_width == width => {
+                let max = max.checked_mul(factor as u64)?;
```

Comment:
Integer conversions a tricky and `as` can be a footgun; let's future proof:

```suggestion
                let max = max.checked_mul(u64::from(factor))?;
```

Reply:
Again: is it better to return `None` than to do a saturating multiplication? I honestly don't know. But if we are using `None` for "any value" then this essentially means that we have two representations for "any value": `None` and `Some(Fact::ValueMax { max: u64::MAX })`.

Reply:
On this topic: it might be worth drawing out a diagram of our lattice in the docs for `Fact` or something and a note about what our representation is for the things where it isn't obvious (top/bottom). Something like:

```
///                None
///               /     \
///              /       \
///             /         \
///     Some(ValueMax)  Some(PointsTo)
///             \         /
///              \       /
///               \     /
///            Err(PccError)
///
/// * The top of the lattice, representing any value, is `None`.
///
/// * The bottom of the lattice, representing the lack of any (valid)
///   value, is an error.
///
/// * The intermediate values are either a maximum bound on the value,
///   or that the value is a pointer to a memory region. Various
///   arithmetic operations are valid on various combinations of these
///   facts. Adding two `MaxValue`s will produce a new `MaxValue` with
///   a maximum of the sum of the parts' maximums. Adding a `MaxValue`
///   to a `PointsTo` will produce a new `PointsTo` that is pointing
///   to a smaller memory region. Etc...
```

Although, now that I think about it, I don't think this is even really a lattic? Maybe a semi-lattice. I don't think we have both a bottom and a top... I guess the real bottom would be "no value" for dead code, but we don't really have that given that we aren't propagating facts ourselves, merely checking them.

And I guess I would expect `None` to represent "no value" rather than "any value" so maybe that is why I am hung up on this `None` stuff.

I don't know, again this comment is all over the place and rambly and doesn't have a real concrete suggestion.

What do you think?

Reply:
Done; this method will return a `Some(..)` in any case we have a `ValueMax`, and at worst return the range of the given bitwidth.

Re: `None` -- I guess the way I think about it is that "we have no fact" (in the same way that the facts attached to values are an array of `Option<Fact>`) -- we just can't say/derive anything. I'm not sure I want to describe this in terms of a (semi-)lattice as we don't have a meet-operation anywhere, and that's pretty intentional -- we aren't doing any fixpoint analysis or inference, we just check what's given. (I guess subsumption technically induces a partial order and a least-upper-bound might exist, but...)